### PR TITLE
RSDK-3414: Implement timeout for MoveOnMap

### PR DESCRIPTION
**Pull request by nthakur7**
Added timeout within `GoToInputs()` so that if neither position nor heading of the base have changed in 10 seconds then a timeout error will be returned. Information about change in the base's pose is sent on the `positionChanged` channel.

### CODE REVIEW

File: components/base/kinematicbase/kinematics.go [hunk 1]
```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
...
 	// when the base is within the positional threshold of the goal, exit the loop
-	for err = ctx.Err(); err == nil; err = ctx.Err() {
-		col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
-		if err != nil {
-			return err
-		}
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	go func() {
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:128]: This is a no-no in production code. If something in your goroutine panicked, it would crash all of RDK.

To solve this we have this function which you can pass this codeblock directly into: https://github.com/viamrobotics/goutils/blob/b45e9e59763295615c25debe7861f69b55d0f52a/runtime.go#L141

File: components/base/kinematicbase/kinematics.go [hunk 2]
```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
...
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	utils.PanicCapturingGo(func() {
+		prevDistErr := -1
+		prevHeadingErr := math.Inf(-1)
+		var contextErr error
+		for contextErr = ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
+			if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
+				poseChange <- true
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:131]: What you have here is a bit more complex than it needs to be, and may be prone to lockups if something goes wrong.

I would recommend removing the `poseChange` channel, the `time.NewTimer(timeout)`, and the entirety of the other goroutine which watches the `poseChange` channel.

All of that removed code can become just the following:
```
lastUpdate := time.Now()
for contextErr = ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
	if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
		lastUpdate = time.Now()
	} else if time.Since(lastUpdate) > timeout {
		throw an error
	}
```

File: components/base/kinematicbase/kinematics.go [hunk 3]
```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
...
+				return
+			}
+			if !col {
+				movementErr <- errors.New("base has deviated too far from path")
+				return
+			}
+
+			// get to the x, y location first - note that from the base's perspective +y is forward
+			desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
+			prevDistErr = distErr
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:146]: This seems like this can cause problems.

If you're iterating this loop rapidly, you may be sampling faster than the base can actually travel `distEpsilon` mm, in which case `poseChanged` will always be false.

I think you only want to update prevDistErr and prevHeadingErr if `poseChanged` returned true

File: components/base/kinematicbase/kinematics.go [hunk 4]
```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
...
+			prevHeadingErr = headingErr
+			distErr, headingErr, err = ddk.errorState(current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:148]: I feel like naming these with "error" is a bit misleading, since any base movement will have a high amount of "error" for any motion, unrelated to how perfectly it executes.

File: components/base/kinematicbase/kinematics.go (hunk 2 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:131]: Should I keep the goroutine that the whole for loop exists in? I created that so it could run concurrently with the other goroutine which was watching for a pose change but it seems like it's no longer necessary?

Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:131]: Good point, there's probably not a reason to use goroutines at all in that case.

File: components/base/kinematicbase/kinematics.go (hunk 4 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:148]: the "error" is referring to the error between where the base is and where the goal is. I named it after the fact that `errorState` says it returns distance error and heading error. Not really sure what a better name for this could be

Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:148]: I understand that; it's that it's odd to refer to the starting point as an "error state".

Naively I would expect "error" to refer to, say, the deviation from an intended path. Not the state of simply not being at the goal. "Distance" might be a better term here.

File: components/base/kinematicbase/kinematics.go (hunk 2 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:131]: talked to @raybjork - looks like the `issueCommand` calls will block until they can successfully complete and this new structure won't be able to interrupt that, which is why the separate goroutine is necessary for the timer

File: components/base/kinematicbase/kinematics.go (hunk 4 above)
Comment (raybjork) [reply components/base/kinematicbase/kinematics.go:148]: Disagree Peter, we're writing a control loop here and in control theory is the difference between the desired and the current.

Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:148]: This is just a disagreement over what is "desired".

For a control loop that follows a path, I would expect "desired" to be "the optimal path to the goal". As written it's the goal state itself, not a path to it.

File: components/base/kinematicbase/kinematics.go (hunk 2 above)
Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:131]: Got it. That makes sense.

In that case the correct structure is still not what we have here, but rather what we discussed above with no goroutines, but with the addition of a timeout to the context `context.WithTimeout(ctx, timeout)` each time we have a position update.

File: components/base/kinematicbase/kinematics.go [hunk 5]
```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
...
-			return err
+			case <-t.C: // timer expired before a position change occurred
+				movementErr <- errors.New("movement timeout")
+			}
 		}
-	}
-	return err
+	})
+	goToInputsErr := <-movementErr // waits on error value to be returned from either timer or movement code
+	return goToInputsErr
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:203]: If there's a non-nil error, we would presumably want to stop the base, right?

Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:203]: I believe this will be checked in MoveOnMap and the movement This is how the structure has always been

Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:203]: Good with this then

File: components/base/kinematicbase/kinematics.go [hunk 6]
```diff
@@ -106,42 +111,86 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
...
+			if err != nil {
+				movementErr <- err
 			}
 		}
+		movementErr <- ctx.Err()
+	})
 
-		current, err = ddk.CurrentInputs(ctx)
+	// goroutine for watching for movement timeout
+	utils.PanicCapturingGo(func() {
```
Comment (nthakur7) [inline components/base/kinematicbase/kinematics.go:159]: this goroutine has a lot of repetitive code with calculating errorState and CurrentInputs but I wasn't sure of a better way to do this and still ensure that prevDistErr and prevHeadingErr get initialized at some point

File: components/base/kinematicbase/kinematics.go [hunk 7]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
-		// get to the x, y location first - note that from the base's perspective +y is forward
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+	utils.PanicCapturingGo(func() {
+		// this loop polls the error state and issues a corresponding command to move the base to the objective
+		// when the base is within the positional threshold of the goal, exit the loop
+		for contextErr := ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
```
Comment (raybjork) [inline components/base/kinematicbase/kinematics.go:126]: I think this can stay as `err` as well, no?

File: components/base/kinematicbase/kinematics.go [hunk 8]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
-			} else {
-				return err
+			}
+
+			current, err = ddk.CurrentInputs(cancelContext)
+			if err != nil {
+				movementErr <- err
 			}
 		}
+		movementErr <- ctx.Err()
```
Comment (raybjork) [inline components/base/kinematicbase/kinematics.go:159]: When you change the `contextErr` to `err` we can just do `movementErr <- err` again

File: components/base/kinematicbase/kinematics.go [hunk 9]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
-		if err != nil {
-			return err
+	// goroutine for watching for movement timeout
+	utils.PanicCapturingGo(func() {
+		lastUpdate := time.Now()
+		prevDistErr := math.Inf(-1)
+		prevHeadingErr := math.Inf(-1)
+
+		for {
+			currentPose, err := ddk.CurrentInputs(ctx)
```
Comment (raybjork) [inline components/base/kinematicbase/kinematics.go:169]: This doesn't return a pose so preference to rename to just `current`

File: components/base/kinematicbase/kinematics.go [hunk 10]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
 		}
+		movementErr <- ctx.Err()
+	})
 
-		current, err = ddk.CurrentInputs(ctx)
-		if err != nil {
-			return err
+	// goroutine for watching for movement timeout
+	utils.PanicCapturingGo(func() {
+		lastUpdate := time.Now()
```
Comment (raybjork) [inline components/base/kinematicbase/kinematics.go:164]: I know we probably talked about this but I forget.  Why the change to a time based system over using a timer?

File: components/base/kinematicbase/kinematics.go [hunk 11]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
+		prevDistErr := math.Inf(-1)
+		prevHeadingErr := math.Inf(-1)
+
+		for {
```
Comment (raybjork) [inline components/base/kinematicbase/kinematics.go:168]: Slightly worried that running this without any delay is going to introduce a lot of overhead

File: components/base/kinematicbase/kinematics.go [hunk 12]
```diff
@@ -258,11 +304,17 @@ func (ddk *differentialDriveKinematics) newValidRegionCapsule(starting, desired
...
+		2*deviationThreshold+positionErr,
 		"")
 	if err != nil {
 		return nil, err
 	}
 
 	return capsule, nil
 }
+
+func poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr float64) bool {
```
Comment (raybjork) [inline components/base/kinematicbase/kinematics.go:316]: @biotinker, you've worked with metrics the most, which function should we be using here instead of writing something new?

File: services/motion/builtin/builtin_test.go [hunk 13]
```diff
@@ -230,20 +248,7 @@ func TestMoveOnMap(t *testing.T) {
...
-		f := func() ([]byte, error) {
-			bytesRead, err := file.Read(chunk)
-			if err != nil {
-				defer utils.UncheckedErrorFunc(file.Close)
-				return nil, err
-			}
-			return chunk[:bytesRead], err
-		}
-		return f, nil
+		return getPointCloudMap(path)
```
Comment (raybjork) [inline services/motion/builtin/builtin_test.go:251]: could write this in one line

File: services/motion/builtin/builtin_test.go [hunk 14]
```diff
@@ -310,6 +315,51 @@ func TestMoveOnMap(t *testing.T) {
...
+	myRobot, err := robotimpl.New(ctx, cfg, logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, myRobot.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	injectSlam := inject.NewSLAMService("test_slam")
+	injectSlam.GetPointCloudMapFunc = func(ctx context.Context) (func() ([]byte, error), error) {
+		path := filepath.Clean(artifact.MustPath("pointcloud/octagonspace.pcd"))
+		return getPointCloudMap(path)
```
Comment (raybjork) [inline services/motion/builtin/builtin_test.go:332]: this can be one line too

**Review by raybjork — CHANGES_REQUESTED** [review]

File: components/base/kinematicbase/kinematics.go [hunk 15]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
 func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired []referenceframe.Input) (err error) {
 	// create capsule which defines the valid region for a base to be when driving to desired waypoint
 	// deviationThreshold defines max distance base can be from path without error being thrown
-	current, err := ddk.CurrentInputs(ctx)
-	if err != nil {
-		return err
+	current, inputsErr := ddk.CurrentInputs(ctx)
```
[the remark this answers is not in the document]
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:110]: errors are named differently to avoid the goroutines accessing the same shared err variable and potentially overwriting each other's errors.

File: components/base/kinematicbase/kinematics.go [hunk 16]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
-			return err
-		}
+	utils.PanicCapturingGo(func() {
+		// this loop polls the error state and issues a corresponding command to move the base to the objective
+		// when the base is within the positional threshold of the goal, exit the loop
+		for contextErr := ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
+			utils.SelectContextOrWait(ctx, 100*time.Millisecond)
+			col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
+			if err != nil {
+				movementErr <- err
```
[the remark this answers is not in the document]
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:130]: oh yes I meant to do that just forgot to add it back in

File: components/base/kinematicbase/kinematics.go (hunk 9 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:169]: can't name it `current` because the other goroutine is modifying the `current` variable that was created at the top of the function. but will rename to smth else

File: components/base/kinematicbase/kinematics.go [hunk 17]
```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
-	// when the base is within the positional threshold of the goal, exit the loop
-	for err = ctx.Err(); err == nil; err = ctx.Err() {
-		col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
-		if err != nil {
-			return err
-		}
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	cancelContext, cancel := context.WithCancel(ctx)
```
[the remark this answers is not in the document]
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:121]: chose to use cancel context after discussing w @biotinker

File: components/base/kinematicbase/kinematics.go (hunk 12 above)
Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:325]: This would be a `motionplan.SquaredNormSegmentMetric` with two poses

File: components/base/kinematicbase/kinematics.go [hunk 18]
```diff
@@ -102,50 +107,100 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
-		// get to the x, y location first - note that from the base's perspective +y is forward
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+	utils.PanicCapturingGo(func() {
+		// this loop polls the error state and issues a corresponding command to move the base to the objective
+		// when the base is within the positional threshold of the goal, exit the loop
+		for err := ctx.Err(); err == nil; err = ctx.Err() {
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:126]: This should use `cancelContext`.

Since `cancelContext` was created with `ctx`, it will error when `ctx` errors, or if you cancel it.

File: components/base/kinematicbase/kinematics.go [hunk 19]
```diff
@@ -102,50 +107,100 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
 			}
 		}
+		movementErr <- err
+	})
 
-		current, err = ddk.CurrentInputs(ctx)
-		if err != nil {
-			return err
+	// goroutine for watching for movement timeout
+	utils.PanicCapturingGo(func() {
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:169]: This does not need to be in a goroutine; nothing here is blocking.

File: components/base/kinematicbase/kinematics.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:121]: A cancellable context is what allows us to actually cease the movement when we detect and return an error.

This should have a `defer cancel()` so that any exit of this function will cause the spawned goroutines to also cancel.

File: components/base/kinematicbase/kinematics.go [hunk 20]
```diff
@@ -102,50 +107,100 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
+			if err != nil {
+				movementErr <- err
+				return
+			}
+			if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
+				lastUpdate = time.Now()
+				prevDistErr = distErr
+				prevHeadingErr = headingErr
+			} else if time.Since(lastUpdate) > timeout {
+				cancel()
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:191]: If the `defer cancel()` is added above this can go away

**Review by biotinker — CHANGES_REQUESTED** [review]
Looking better, but there's still a chance for lockups with the goroutines as designed. Let's remove the second goroutine.

File: components/base/kinematicbase/kinematics.go (hunk 19 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:169]: how should the movementErr be checked for a value in this case? the only way I can think of is checking if it contains a value on every loop but according  to online stuff it looks like checking the length of a channel and then taking a value from it isn't best practice

Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go]: I pushed an updated version without the second goroutine so you can see how I implemented it but again I'm not fully sure this is the right way to do it

File: components/base/kinematicbase/kinematics.go [hunk 21]
```diff
@@ -102,50 +110,96 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
+
+	// watching for movement timeout
+	lastUpdate := time.Now()
+	prevDistErr := math.Inf(-1)
+	prevHeadingErr := math.Inf(-1)
 
-		current, err = ddk.CurrentInputs(ctx)
+	for {
+		utils.SelectContextOrWait(ctx, 100*time.Millisecond)
+		if len(movementErr) > 0 {
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:179]: You are correct in your earlier comment that checking the length of a channel is not best practice. This should be:
```
select {
	case err := <- movementErr
		return err
	default:
}
```

File: components/base/kinematicbase/kinematics.go (hunk 12 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:324]: this requires converting an Input array to a Pose every time a new currentInputs is calculated because this function requires having the position in Pose form. is that fine?

Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:324]: Actually; I take it back. We should be able to use a `motionplan.L2InputMetric` directly here, not a `SquaredNormSegmentMetric`.

For future reference though; the construction of a `Pose` from a point + orientation as we would be doing here, is not particularly computationally expensive and would be fine to do at the cadence of this code.

File: components/base/kinematicbase/kinematics.go (hunk 10 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go]: don't need second goroutine now because nothing is blocking

File: components/base/kinematicbase/kinematics.go [hunk 22]
```diff
@@ -102,50 +110,95 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+	utils.PanicCapturingGo(func() {
+		// this loop polls the error state and issues a corresponding command to move the base to the objective
+		// when the base is within the positional threshold of the goal, exit the loop
+		for err := cancelContext.Err(); err == nil; err = cancelContext.Err() {
+			utils.SelectContextOrWait(ctx, 100*time.Millisecond)
```
Comment (raybjork) [inline components/base/kinematicbase/kinematics.go:131]: [q] Do we still need to have this? Would be good to not have to sleep in the main control loop

**Review by raybjork — APPROVED** [review]
Just one question but otherwise looks good!

File: components/base/kinematicbase/kinematics.go (hunk 22 above)
Comment (biotinker) [reply components/base/kinematicbase/kinematics.go:131]: Without some wait here this will peg a CPU at 100%.

It could be shorter than 100ms though

File: components/base/kinematicbase/kinematics.go [hunk 23]
```diff
@@ -102,50 +110,95 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
...
+				movementErr <- err
+				return
 			}
 		}
+		movementErr <- err
+	})
+
+	// watching for movement timeout
+	lastUpdate := time.Now()
+	var prevInputs []referenceframe.Input
```
Comment (biotinker) [inline components/base/kinematicbase/kinematics.go:174]: This has the ability to cause a nil pointer dereference on the first iteration; the only reason it doesn't currently is a coincidence of the L2InputMetric implementation.

This should be initialized to CurrentInputs rather than to nil

File: components/base/kinematicbase/kinematics.go (hunk 22 above)
Comment (nthakur7) [reply components/base/kinematicbase/kinematics.go:131]: I'm not really sure when this is or isn't necessary

**Review by biotinker — APPROVED** [review]
LGTM. Nice work with this.
